### PR TITLE
System db restore

### LINKED ISSUE #964: New Command: Restore-DbaSystemDatabase for 1.0
### Is this a *feature* OR *bug*:

feature

### Details

Hey @Stuart-Moore - would you be able to work on this? 

Here are some tutorials:

https://msdn.microsoft.com/en-us/library/ms345408%28SQL.90%29.aspx
https://msdn.microsoft.com/en-us/library/ms345408(v=sql.130).aspx

I tried once and then ran out of time because I got caught up trying to offer what was essentially Move-DbaSystemDatabase (which was hell).

Here's what I had when I said F it. I couldn't get it to work consistently enough but I think it only was annoying for the move and not the restore.

Please ignore the bad practices -- this is from 3 years ago.

```
Function Restore-SQLSystemDatabase {
        <#
            .SYNOPSIS
            Restores master, model and msdb.
            .EXAMPLE
			 Restore-SQLSystemDatabase $servername $dbname $backupfile $systemdir
            .OUTPUTS
                $true if success
                $true if failure
        #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
            [ValidateNotNullOrEmpty()]
			[object]$servername,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$dbname,

			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$backupfile,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$systemdir
			
        )

	$restore = New-Object "Microsoft.SqlServer.Management.Smo.Restore"
	$restore.ReplaceDatabase = $true

	switch ($dbname) {
		"master" { 	$datalogical = "master"; $dataphysical = "$systemdir\master.mdf"
					$loglogical = "mastlog"; $logphysical = "$systemdir\mastlog.ldf" }
		"model" { 	$datalogical = "modeldev"; $dataphysical = "$systemdir\model.mdf"
					$loglogical = "modellog"; $logphysical = "$systemdir\modellog.ldf" }
		"msdb" { 	$datalogical = "MSDBData"; $dataphysical = "$systemdir\MSDBData.mdf"
					$loglogical = "MSDBLog"; $logphysical = "$systemdir\MSDBLog.ldf" }
	}
	
	$movedatafile = New-Object "Microsoft.SqlServer.Management.Smo.RelocateFile" 
	$movedatafile.LogicalFileName = $datalogical
	$movedatafile.PhysicalFileName = $dataphysical
	$null = $restore.RelocateFiles.Add($movedatafile)

	$movelogfile = New-Object "Microsoft.SqlServer.Management.Smo.RelocateFile" 
	$movelogfile.LogicalFileName = $loglogical
	$movelogfile.PhysicalFileName = $logphysical
	$null = $restore.RelocateFiles.Add($movelogfile)
	
	try {
		$restore.ReplaceDatabase = $true
		$restore.Database = $dbname
		$restore.Action = "Database"
		$restore.NoRecovery = $false
		$device = New-Object -TypeName Microsoft.SqlServer.Management.Smo.BackupDeviceItem
		$device.name = $backupfile
		$device.devicetype = "File"
		$restore.Devices.Add($device)
		$restore.sqlrestore($servername)
		return $true
	} catch { return $false }
}

Function Restore-SQLMasterDB {
 <#
	.SYNOPSIS
		Restoring the master database requires starting up in a specific mode. This function changes the startup params, 
		restarts sql, and restores the database.
	.EXAMPLE
		$result = Restore-SQLMasterDB $server $sqlservice $backupfile $startparams
	.OUTPUTS
		true or false
			
 #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [object]$server,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [object]$sqlservice,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$master,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$startparams

		)

	$systemdir = $server.MasterDBPath
	$netbiosname = Get-NetBIOSName $server
	$servername = $server.name
	$instance = ($servername.split("\"))[1]
	if ($instance -eq $null) { $instance = "MSSQLSERVER" }
	
	Write-Host "Changing startup parameters to allow restore of master db." -ForegroundColor Yellow
	$sqlservice.StartupParameters = "-m;$($startparams)"; $sqlservice.Alter()

	Write-Warning "Attempting to restart SQL Services.."
	if ($server.isclustered) {
		try {
				$remotesqlservice = Get-Service -ComputerName $netbiosname | Where-Object { $_.DisplayName -eq "SQL Server ($instance)" }
				$clusteredservices = Get-ClusterResource -Cluster $servername | Where-Object { $_.Name.StartsWith("SQL Server") -and $_.OwnerGroup -eq "SQL Server ($instance)" }
				$clusteredservices | Stop-ClusterResource
				try { $remotesqlservice | Start-Service } catch { throw "bullshit" }
		} catch { 
			Write-Host "SQL Service is clustered and Service Control Manager could not be contacted. Quitting." -ForegroundColor Red
			return $false
		}
	} else {
		Stop-SQLService $sqlservice
		Start-SQLservice $sqlservice
	}
	
	$sqlservice.StartupParameters = "-T3608;-m;$($startparams)"
	$sqlservice.Alter()
	
	Write-Host "Restoring master" -ForegroundColor Yellow
	$result = Restore-SQLSystemDatabase $servername "master" $master $systemdir
	return $result
}

Function Start-ReplaceSystemDBs {
 <#
	.SYNOPSIS
		Performs all of the steps needed to restore master, msdb and model.
	.EXAMPLE
		Start-ReplaceSystemDBs $server $RestoreFromDirectory
	.OUTPUTS
		true or false
 #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [object]$server,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [string]$RestoreFromDirectory
		)
	
	$servername = $server.name
	$netbiosname = Get-NetBIOSName $server
	$instance = ($servername.split("\"))[1]
	if ($instance -eq $null) { $instance = "MSSQLSERVER" }
	
	$server.ConnectionContext.StatementTimeout = 0
	
	$master = Get-ChildItem "$RestoreFromDirectory\master\FULL\*.bak" | sort LastWriteTime | select -last 1
	$msdb = Get-ChildItem "$RestoreFromDirectory\msdb\FULL\*.bak" | sort LastWriteTime | select -last 1
	$model = Get-ChildItem "$RestoreFromDirectory\model\FULL\*.bak" | sort LastWriteTime | select -last 1
	
	Write-Host "Getting master backup version" -ForegroundColor Yellow
	$restore = New-Object "Microsoft.SqlServer.Management.Smo.Restore"
	$device = New-Object -TypeName Microsoft.SqlServer.Management.Smo.BackupDeviceItem $master, "FILE"
	$restore.Devices.Add($device)
	$backupinfo = $restore.ReadBackupHeader($server)
	$backupversion = [version]("$($backupinfo.SoftwareVersionMajor).$($backupinfo.SoftwareVersionMinor).$($backupinfo.SoftwareVersionBuild)")

	if ($backupversion -ne $server.version) {
			Write-Warning "Not the same version (backup: $backupversion, server: $($server.version)). Cannot restore system dbs."
			return $false
	}
	
	if ($server.isclustered) {
		try {
			$remotesqlservice = Get-Service -ComputerName $netbiosname | Where-Object { $_.DisplayName -eq "SQL Server ($instance)" }
			$clusteredservices = Get-ClusterResource -Cluster $servername | Where-Object { $_.Name.StartsWith("SQL Server") -and $_.OwnerGroup -eq "SQL Server ($instance)" }
		} catch { 
			Write-Host "SQL Service is clustered and Service Control Manager could not be contacted. Quitting." -ForegroundColor Red
			return $false
		}
	}
	
	Write-Host "Master backup version and SQL Server match. Going forward." -ForegroundColor Green
	
	$null = [Reflection.assembly]::LoadWithPartialName("Microsoft.SqlServer.SqlWmiManagement")

	$sqlserver = New-Object -TypeName Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $netbiosname
	$systemdir = $server.MasterDBPath
	$sqlservice = ($sqlserver.Services | Where-Object { $_.DisplayName -eq "SQL Server ($instance)" })
	$sqlagent = ($sqlserver.Services | Where-Object { $_.DisplayName -eq "SQL Server Agent ($instance)" })
	
	Write-Host "Backing up current master just in case.." -ForegroundColor Yellow
	$currentmasterbackup = "$RestoreFromDirectory\master\FULL\ps-migrationmaster.bak"
	$sql = "BACKUP DATABASE [master] TO  DISK = N'$currentmasterbackup' WITH NOFORMAT, INIT"
	try { 
		$server.databases['master'].ExecuteNonQuery($sql)
		Write-Host "Successfully backed up master database" -ForegroundColor Green
	} catch {
		Write-Warning "Couldn't backup master. Aborting."
		return $false
	}
	
	$startparams = $sqlservice.StartupParameters
	if ($servername -eq 'sqlcluster') {
		$startparams = "-dM:\MSSQL12.MSSQLSERVER\MSSQL\DATA\master.mdf;-eM:\MSSQL12.MSSQLSERVER\MSSQL\Log\ERRORLOG;-lM:\MSSQL12.MSSQLSERVER\MSSQL\DATA\mastlog.ldf"
	}
	$result = Restore-SQLMasterDB $server $sqlservice $master $startparams
	
	# IF RESTORE OF MASTER WENT SUCCESSFULLY
	if ($result) {
		$sql = "ALTER DATABASE msdb MODIFY FILE (NAME=MSDBData, FILENAME= '$systemdir\MSDBData.mdf');
			ALTER DATABASE msdb MODIFY FILE (NAME=MSDBLog, FILENAME= '$systemdir\MSDBLog.ldf');
			ALTER DATABASE model MODIFY FILE (NAME=modeldev, FILENAME= '$systemdir\model.mdf');
			ALTER DATABASE model MODIFY FILE (NAME=modellog, FILENAME= '$systemdir\modellog.ldf');
			ALTER DATABASE tempdb MODIFY FILE (NAME=tempdev, FILENAME= '$systemdir\tempdb.mdf');
			ALTER DATABASE tempdb MODIFY FILE (NAME=templog, FILENAME= '$systemdir\templog.ldf');"
		Write-Host "Master successfully restored! Restarting SQL Server." -ForegroundColor Green
		Start-Sleep 5
	} else {
		Write-Host "Master could not be master restored." -ForegroundColor Red
		Write-Host "Aborting system db restores and starting SQL Server in normal mode." -ForegroundColor Red
		$sqlservice.StartupParameters = $startparams
		try { Stop-SQLservice $sqlservice; Start-SQLservice $sqlservice } catch {}
		return $false
}
	
	Write-Host "Resetting startup to normal, and updating msdb, model and tempdb locations." -ForegroundColor Yellow
	$timeout = new-timespan -Minutes 2
	$sw = [diagnostics.stopwatch]::StartNew()
	
	do {	
		try {
			$server.ConnectionContext.Connect()
			$server.databases['master'].ExecuteNonQuery($sql)
			Write-Host "Master successfully updated with new system db locations! Restarting SQL Server." -ForegroundColor Green
			$success = $true
		} catch { 
			write-warning "Connect failed, retrying.."; 
			if ($server.isclustered) {
				if ($remotesqlservice.Status -ne "Stopped") { Stop-SQLservice $remotesqlservice $true }
				Start-SQLservice $remotesqlservice $true 
			} else {
				if ($sqlservice.ServiceState -ne "Stopped") { Stop-SQLservice $sqlservice }
				Start-SQLservice $sqlservice
			}
		}
	} until ($success -eq $true -or $sw.elapsed -gt $timeout) 
	
	$sqlservice.StartupParameters = $startparams; $sqlservice.Alter()
	
	if ($server.isclustered) {
		Stop-SQLService $remotesqlservice $true
	} else { Stop-SQLservice $sqlservice }
	
	Start-SQLservice $sqlservice
	Write-Warning "Restoring msdb"
	Restore-SQLSystemDatabase $server "msdb" $msdb $systemdir
	Write-Warning "Restoring model"
	Restore-SQLSystemDatabase $server "model" $model $systemdir
	$sqlagent.start()

	try { $server.databases.refresh()
			foreach ($db in $server.databases) {
				if ($db.status -ne "Normal") { 
					try { $dbname = $db.name
						$sql = "ALTER DATABASE [$dbname] SET EMERGENCY; drop database [$dbname]"
						$server.databases['master'].ExecuteNonQuery($sql)
					} catch {}
				}
			} 
	} catch {}
	Remove-Item $currentmasterbackup
	Write-Host "Systems databases successfully restored! Moving on to user databases.." -ForegroundColor Green
	return $true
}

Function Start-SQLService {
 <#
	.SYNOPSIS
	Restarting SQL Server via SQL's WMI objects returns $true too quickly. This function waits until the SQL service has started or a timeout occurs.
	.EXAMPLE
	$result = Start-SQLService $sqlservice $clustered
	
	.OUTPUTS
	  true or false
			
 #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [object]$service,
			
			[Parameter(Mandatory = $false)]
            [bool]$clustered	
		)

	$timeout = new-timespan -Seconds 30
	$sw = [diagnostics.stopwatch]::StartNew()

	if ($clustered) {
		do {
			$service.Refresh()
			try { 
				$service.Start() 
				Write-Warning "Starting SQL.." } 
			catch { Write-Warning "Having trouble starting the SQL service. Let's try again." }
			
			Start-Sleep 2
		} 
		until ($service.Status -eq "Running" -or $sw.elapsed -gt $timeout) 
		$status = $service.Status	
	} else {
		do {
			$service.Refresh()
			try { 
				$service.Start()
				Write-Warning "Starting SQL.." } 
			catch { 
				Write-Warning "Having trouble starting the SQL service. Let's try again." }
				Start-Sleep 2
		} 
		until ($service.ServiceState -eq "Running" -or $sw.elapsed -gt $timeout) 
		$status = $service.ServiceState
	}
	
	if ($status -eq "Running") { return $true }
	else { return $false }
}

Function Stop-SQLService {
 <#
	.SYNOPSIS
	Restarting SQL Server via SQL's WMI objects returns $true too quickly. This function waits until the SQL service has stopped or a timeout occurs.
	.EXAMPLE
	$result = Stop-SQLService $sqlservice
	
	.OUTPUTS
	  true or false
			
 #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
            [object]$service,
			
			[Parameter(Mandatory = $false)]
            [bool]$clustered	
		)

	$timeout = new-timespan -Seconds 30
	$sw = [diagnostics.stopwatch]::StartNew()

	if ($clustered) {
		do {
			$service.Refresh()
			try {
				$service.Stop() 
				Write-Warning "Stopping SQL.."
			} catch { Write-Warning "Having trouble stopping the SQL service. Let's try again." }
			
			Start-Sleep 2
		} 
		until ($service.Status -eq "Stopped" -or $sw.elapsed -gt $timeout) 
		$status = $service.Status	
	} else {
		do {
			$service.Refresh()
			try { 
				$service.Stop() 
				Write-Warning "Stopping SQL.."
			} catch { Write-Warning "Having trouble stopping the SQL service. Let's try again." }
			
			Start-Sleep 2
		} 
		until ($service.ServiceState -eq "Stopped" -or $sw.elapsed -gt $timeout) 
		$status = $service.ServiceState
	}
	
	if ($status -eq "Stopped") { return $true }
	else { return $false }
}

Function Drop-SQLDatabase {
 <#
            .SYNOPSIS
             Uses SMO's KillDatabase to drop all user connections then drop a database. $server is
			 an SMO server object.
            .EXAMPLE
              Drop-SQLDatabase $server $dbname
            .OUTPUTS
                $true if success
                $false if failure
			
        #>
		[CmdletBinding()]
        param(
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
			[object]$server,
			
			[Parameter(Mandatory = $true)]
			[ValidateNotNullOrEmpty()]
			[string]$dbname
		)
		
	try {
		$server.KillDatabase($dbname)
		$server.refresh()
		Write-Host "Successfully dropped $dbname on $($server.name)." -ForegroundColor Green
		return $true
	}
	catch {	return $false }
}
```

### PR DESCRIPTION
**Pull request by Stuart-Moore**
Fixes #964 

Changes proposed in this pull request:
 -  Adds the option to use the restore-dbadatabase work to restore system dbs
 - Wrapper function to do all the restarts and parameters changers Restore-DbaSystemDatabase

Notes:
 this work 90% of the time with clustered instances, but  the cluster occasionally sticks it's nose in and really stuffs things up. Only surefire way to do it is to use the FCI module, but that's not something we want as a dependancy. So it warns and stops if  pointed at a clustered instance and the user hasn't indicated they'll accept the risks.
@FriedrichWeinmann  - pulled you in in case you've any better stop/start code in the pipeline than the internal functions

As this is something that could really stuff up someone's system if it goes wrong, would like a review rather than just me merging it, You've all met my tpyos :)

How to test this code: 
 -  [ ] Restore-DbaSystemDatabase -SqlServer localhost\sqlexpress2016 -BackupPath c:\dbatools\master_201705290947.bak -Master 
 -  [ ] Restore-DbaSystemDatabase -SqlServer localhost\sqlexpress2016 -BackupPath c:\dbatools\master_201705290947.bak -Master -Confirm:$false
 - [ ]Restore-DbaSystemDatabase -SqlServer localhost\sqlexpress2016 -BackupPath c:\dbatools\sysdbs\ -Master -Model -Msdb

Has been tested on minimum requirements:
- [ ]  Powershell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [ ]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [ ] Working/useful help content, including link to command on dbatools web site
- [x] All examples work as advertised
- [x] Does not contain template content
- [x] Does not contain excessive/unnecessary amounts of comments
- [x] Works remotely
- [x] Works locally (needs to be run as admin due to the underlying WMI implementation)
- [x] Works on lower versions or throws error specifying version not supported
- [x] Works with named instances
- [ ] Works with clustered instances - mostly, see notes
- [ ] Handles offline/read only databases 
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

### CODE REVIEW

**Comment by potatoqualitee** [comment]
😍 !! I'll fix the conflict

**Comment by Stuart-Moore** [comment]
Thanks. That explains why I just got a strange github error while fixing it online!

Am out for a couple of hours now, but will pick up any comments when I get back.

Think this one may need some warning signs on it, as I don't really fancy supporting someone who's used it incorrectly and hosed master (or model, which is more of a pain to repair as I've discovered 🤕 )

**Comment by potatoqualitee** [comment]
hey stuart - please check out https://github.com/sqlcollaborative/dbatools/pull/1354/commits/8764eab61ae86477792ba967342fc42b40524259 to see the changes I made

**Comment by potatoqualitee** [comment]
LOL, I just realized I literally had a dream about this command last night 💤  Probably because I fell asleep hoping it'd be ready for SQLGrillen!

**Comment by Stuart-Moore** [comment]
Fred's got a list of comments, so hopefully I can crack through those quickly and we might have if for friday.

File: functions/Restore-DbaSystemDatabase.ps1 [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+                        Stop-Function -Message "SQL Server not starting nicely, tried to fix, but not responding" 
+                        return
+                    }
+                }
+                if ($server.connectionContext.IsOpen -eq $false) {
+                    $server.connectionContext.Connect()
+                }
+                Write-Message -Level Verbose  -Message  "Beginning Restore of Master"
+                if ($pscmdlet.ShouldProcess("Restoring Master on $ServerName")) {
+                    $RestoreResult += Restore-DbaDatabase -SqlInstance $server -Path $BackupPath -WithReplace -DatabaseFilter master -RestoreTime $RestoreTime -ReuseSourceFolderStructure -SystemRestore
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaSystemDatabase.ps1]: Should failure here interrupt execution?
If yes, you may want to add the `-Silent` parameter, so it throws and the catch block triggers

File: functions/Restore-DbaSystemDatabase.ps1 [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            Write-Message -Level Warning  -Message "An error has occured: $($error[0].Exception.Message)"
+        }
+        finally {
+            if ((Get-DbaSqlService -ComputerName $ServerName -Credential $Credential -Type Engine | Where-Object {$_.DisplayName -like "*$InstanceName*"}).State -ne 'Running') {
+                Start-DbaService -SqlInstance $server -service SqlServer | out-null
+            }
+            Write-Message -Level Verbose  -Message "Resetting Startup Parameters"
+            Set-DbaStartupParameter -SqlInstance $server -StartUpConfig $CurrentStartup  | out-null
+            Stop-DbaService -SqlInstance $server -Service SqlServer | out-null
+            Start-DbaService -SqlInstance $server -service SqlServer | out-null
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaSystemDatabase.ps1]: Thrice more: Credentials ;)

File: functions/Restore-DbaSystemDatabase.ps1 [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            if ($RestartAgent -eq $True) {
+                Write-Message -Level Verbose  -Message "SQL Agent was running at start, so restarting"
+                Start-DbaService -SqlInstance $server -service SqlAgent | out-null
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaSystemDatabase.ps1]: ...

File: functions/Restore-DbaSystemDatabase.ps1 [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+            }
+            if ($RestartFullText -eq $True) {
+                Write-Message -Level Verbose  -Message "Full Text was running at start, so restarting"
+                Start-DbaService -SqlInstance $server -service FullText | out-null
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaSystemDatabase.ps1]: ...

**Comment by Stuart-Moore** [comment]
@FriedrichWeinmann  - thanks, I'll start working my way through those

@potatoqualitee  - don't think you'll be using this at grillen, I can't type that fast 😉

**Comment by potatoqualitee** [comment]
hahah alright, ill let them know it's in the works tho! 💓

**Comment by potatoqualitee** [comment]
aw shit, we have conflicts :( @ConstantineK - can you do the cherry pick or should we change something?

**Comment by Stuart-Moore** [comment]
There'll be loads as it's ancient. And I've not had a chance to finish all
of Fred's requests yet.

**Comment by ConstantineK** [comment]
Let me know if/when you are ready @Stuart-Moore I would be glad to start pulling apart the merge conflicts :D

**Comment by potatoqualitee** [comment]
I can also put it out for beta release on Wednesday if it is ready 👼

**Comment by gbargsley** [comment]
I see a lot of discussion on this PR. 

What if anything needs to be tested for this PR?

**Comment by potatoqualitee** [comment]
hey garry, we'll need to come together as a team to resolve the conflicts first - then I'd hurry and merge and then we can test and if it's not ready, just not add it to psd1.

This command will restore master and msdb! How frikken awesome.

**Comment by gbargsley** [comment]
👍

**Comment by potatoqualitee** [comment]
So sorry - this has been opened since March 30th so I'm going to close it. This command will be a fan favorite but will need to be redone in another PR.

**Pull request closed without merging**